base/ltkeys.dtx

```diff
@@ -0,0 +1,454 @@
+%     \item \texttt{.store} --- stores a value in a macro
+%     \item \texttt{.usage} -- defines whether the option can be given only
+%       when loading (\texttt{load}), in the preamble (\texttt{preamble}) or
+%       has no limitation on scope (\texttt{general})
+%   \end{itemize}
+%   The part of the \meta{key} before the \meta{property} is the \meta{name},
+%   with the \meta{value} working with the \meta{property} to define the
+%   behaviour of the option.
+%
+%   For example, with
+%   \begin{verbatim}
+%     \DeclareKeys[mypkg]
+%       {
+%         draft.if          = @mypkg@draft      ,
+%         draft.usage       = preamble          ,
+%         name.store        = \@mypkg@name      ,
+%         name.usage        = load              ,
+%         second-name.store = \@mypkg@other@name
+%       }
+%   \end{verbatim}
+%   three options would be create. The option \texttt{draft} can be given
+%   anywhere in the preamble, and will set a switch called \cs{if@mypkg@draft}.
+%   The option \texttt{name} can only be given during package loading, and will
+%   save whatever value it is given in \cs{@mypkg@name}. Finally, the option
+%   \texttt{second-name} can be given anywhere, and will save its value in
+%   \cs{@mypkg@other@name}.
+%
+%   Keys created \emph{before} the use of
+%   \cs{ProcessKeyOptions}/\cs{ProcessKeyPackageOptions} act as package options.
+% \end{function}
+%
+% \begin{function}{\ProcessKeyOptions}
+%   \begin{syntax}
+%     \cs{ProcessKeyOptions} \oarg{family}
+%   \end{syntax}
+%   The \cs{ProcessKeyOptions} function is used to check the current
+%   option list against the keys defined for \meta{family}. Global (class)
+%   options and local (package) options are checked when this function
+%   is called in a package.
+% \end{function}
+%
+% \begin{function}{\ProcessKeyPackageOptions}
+%   \begin{syntax}
+%     \cs{ProcessKeyPackageOptions} \oarg{family}
+%   \end{syntax}
+%   This function works in a similar manner to \cs{ProcessKeyOptions}.
+%   When used in a package, \cs{ProcessKeyPackageOptions}
+%   will not examine any global (class) class options available. In contrast,
+%   \cs{ProcessKeyOptions} does parse class options (in common with the
+%   classical \cs{ProcessOptions} command).
+% \end{function}
+%
+% \begin{function}{\SetKeys}
+%   \begin{syntax}
+%     \cs{SetKeys} \oarg{family} \Arg{keyvals}
+%   \end{syntax}
+%   Sets (applies) the explicit list of \meta{keyvals}  for the \meta{family}:
+%   it the latter is not given, the value of \cs{@currname} used. This command
+%   may be used within a package to set options before or after using
+%   \cs{ProcessKeyOptions}/\cs{ProcessKeyPackageOptions}.
+% \end{function}
+%
+% \StopEventually{}
+%
+% \subsection{Implementation of \pkg{ltkeys}}
+%
+%    \begin{macrocode}
+%<@@=keys>
+%    \end{macrocode}
+%
+%    \begin{macrocode}
+%<*2ekernel>
+%    \end{macrocode}
+%
+%    \begin{macrocode}
+\ExplSyntaxOn
+%    \end{macrocode}
+%
```

Comment:
this should of course use the declaremodule method for rollback (but I understand why that is not yet there :-)

Reply:
A policy thing I guess :) I was imagining that as this was strictly opt-in, it didn't need to be rolled back.

Reply:
Even then it should have a rollback and thiss way also a roll forward. For example, when Overleaf is a year behind. Stating that this here is a new module makes this fairly painless.

Reply:
I maintain that making this a rollback module is essential and in line with our policy to support rolling forward, e.g., on overleaf. So I think we should do this. I can have a go at it if you don't want to.